ambry-router/src/test/java/com/github/ambry/router/ParanoidDurabilityOperationTrackerTest.java

```diff
@@ -233,6 +235,55 @@ public void justEnoughRemoteReplicasTest() {
         verifySuccessfulReplicas(responseReplicas, localReplicaSuccessTarget, remoteReplicaSuccessTarget));
   }
 
+  /**
+   * A test with 2 remote replicas, both of which time out. This should result in a successful outcome, triggering the
+   * remote attempt limit exceeded mechanism.
+   */
+  @Test
+  public void timedOutRemoteReplicasTest() {
+    int localReplicaSuccessTarget = 2;
+    int remoteReplicaSuccessTarget = 2;
+    int remoteAttemptLimit = 2;
+    localDcName = new String("dc-0");
+    String remoteDcName1 = new String("remote-1");
+    String remoteDcName2 = new String("remote-2");
+
+    List<Port> portList = Collections.singletonList(new Port(PORT, PortType.PLAINTEXT));
+    List<String> mountPaths = Collections.singletonList("mockMountPath");
+    List<ReplicaSpec> testReplicas = new ArrayList<ReplicaSpec>(Arrays.asList(new ReplicaSpec[] {
+        new ReplicaSpec(localDcName, ReplicaState.LEADER, portList, mountPaths),
+        new ReplicaSpec(localDcName, ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec(localDcName, ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec(remoteDcName1, ReplicaState.STANDBY, portList, mountPaths),
+        new ReplicaSpec(remoteDcName2, ReplicaState.STANDBY, portList, mountPaths)}));
+
+    mockPartition = new MockPartitionId();
+    mockClusterMap = new MockClusterMap(false, buildDataNodeList(testReplicas), 1, Collections.singletonList(mockPartition), localDcName);
+    populateReplicaList(testReplicas);
+
+    responseReplicas.clear();
+    ParanoidDurabilityOperationTracker ot = getParanoidDurabilityOperationTracker(2, 2, localReplicaSuccessTarget, remoteReplicaSuccessTarget, remoteAttemptLimit,  true);
+    while (!ot.hasSucceeded()) {
+      sendRequests(ot, 4);
+      for (int i = 0; i < 4; i++) {
+        ReplicaId currentReplica = inflightReplicas.poll();
+        responseReplicas.add(currentReplica);
+        if (currentReplica.getDataNodeId().getDatacenterName().equals(remoteDcName1) ||
+            currentReplica.getDataNodeId().getDatacenterName().equals(remoteDcName2) ) {
+          ot.onResponse(currentReplica, TrackedRequestFinalState.TIMED_OUT);
+        } else {
+          ot.onResponse(currentReplica, TrackedRequestFinalState.SUCCESS);
+        }
+        //ot.onResponse(currentReplica, TrackedRequestFinalState.SUCCESS);
+      }
+    }
+    assertTrue("Operation should have succeeded", ot.hasSucceeded());
+    assertTrue("Operation should be done", ot.isDone());
+    assertFalse("Operation should not have failed", ot.hasFailed());
```

Comment:
could you also verify that the counter metric is incremented by one?

Reply:
Yeah, added a check to confirm that the metrics is correctly incremented.